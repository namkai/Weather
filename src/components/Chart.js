import c3 from 'c3';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../style/Chart.css';

class Chart extends Component {
	constructor(props) {
		super(props);
		this.state = {
			chartType: 'line',
		};
	}

	componentDidMount() {
		this.updateChart();
	}

	componentDidUpdate() {
		this.updateChart();
	}

	setBarChart = () => {
		this.setState({ chartType: 'bar' });
	};

	setLineChart = () => {
		this.setState({ chartType: 'line' });
	};
	updateChart = () => {
		c3.generate({
			bindto: '#chart',
			data: {
				x: 'x',
				columns: [
					['x', ...this.props.forecast[1]],
					['Temperature', ...this.props.forecast[0]],
				],
				colors: {
					Temperature: 'orange',
				},
				type: this.state.chartType,
			},
			axis: {
				x: {
					type: 'timeseries',
					tick: {
						format: '%d-%m-%Y',
					},
				},
			},
		});
	};
	generateGraphButton = () => {
		if (this.state.chartType === 'line') {
			return (
				<div>
					<a href="#" onClick={this.setBarChart} className="btn btn-outline-primary px-3">Bar Chart</a>
				</div>
			);
		}
		return (
			<div>
				<a href="#" onClick={this.setLineChart} className="btn btn-outline-primary px-3">Line Chart</a>
			</div>
		);
	};

	render() {
		if (this.props.forecast[0].length === 0) {
			return (
				<div/>
			);
		}
		return (
			<div>
				<div className="hr-divider mt-5 mb-3">
					<h3 className="hr-divider-content hr-divider-heading">5 Day Forecast</h3>
				</div>
				<div id="chart"/>
				{this.generateGraphButton()}
			</div>
		);
	}
}

Chart.propTyes = {
	forecast: PropTypes.arrayOf(PropTypes.array)
}

export default Chart;

