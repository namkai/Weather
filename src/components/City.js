import React from 'react';

const City = ({ name }) => {
  return (
	  <li className="nav-item">
		  <a className="nav-link" href="#">{name}</a>
	  </li>
  );
};

export default City;