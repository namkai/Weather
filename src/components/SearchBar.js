import React from 'react';
import { connect } from 'react-redux';
import { fetchWeather, updateQuery } from '../actions';

export const SearchBar = ({ input, updateQuery, fetchWeather }) => (
  <div className="row">
    <div className="col-lg-6">
      <div className="input-group">
        <input
          type="text"
          className="form-control"
          value={input}
          onChange={e => updateQuery(e.target.value)}
          placeholder="Search for..."
        />
        <span className="input-group-btn">
          <button
            className="btn btn-default"
            onClick={() => fetchWeather(input)}
            type="button"
          >Get Weather!
          </button>
        </span>
      </div>
    </div>
  </div>
);

export default connect(({ input }) => ({ input }), { updateQuery, fetchWeather })(SearchBar);
