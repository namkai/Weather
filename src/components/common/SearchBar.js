import PropTypes from 'prop-types';
import React, { Component } from 'react';

class SearchBar extends Component {
	constructor(props) {
		super(props);
		this.state = {
			query: '',
		};
	}

	updateQuery = (query) => {
		this.setState(() => ({ query }));
	};
	searchForWeather = (e) => {
		e.preventDefault();
		this.props.fetchWeather(this.state.query);
		this.setState(() => ({ query: '' }));
	};

	render() {
		return (
			<form onSubmit={this.searchForWeather} className="sidebar-form">
				<input
					placeholder="Search for a City"
					className="form-control"
					value={this.state.query}
					onChange={e => this.updateQuery(e.target.value)}
				/>
				<button type="submit" className="btn-link">
					<span className="icon icon-magnifying-glass"/>
				</button>
			</form>
		);
	}
}
;

SearchBar.defaultProps = {
	input: '',
	updateQuery: PropTypes.func,
	fetchWeather: PropTypes.func,
};

SearchBar.propTypes = {
	input: PropTypes.string,
	updateQuery: PropTypes.func,
	fetchWeather: PropTypes.func,
};

export default SearchBar;

