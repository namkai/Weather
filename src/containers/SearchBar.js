import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions/index';

export const SearchBar = ({ input, updateQuery, fetchWeather }) => {
	const searchForWeather = (e) => {
		e.preventDefault();
		fetchWeather(input);
	};
	return (
    <form onSubmit={searchForWeather} className="input-group">
      <input
        placeholder="Get a five-day forecast in your favorite cities"
        className="form-control"
        value={input}
        onChange={e => updateQuery(e.target.value)}
      />
      <span className="input-group-btn">
        <button type="submit" className="btn btn-secondary">Submit</button>
      </span>
    </form>
	);
};

SearchBar.defaultProps = {
	input: '',
	updateQuery: PropTypes.func,
	fetchWeather: PropTypes.func,
};

SearchBar.propTypes = {
	input: PropTypes.string,
	updateQuery: PropTypes.func,
	fetchWeather: PropTypes.func,
};

export default connect(({ input }) => ({ input }), actions)(SearchBar);
