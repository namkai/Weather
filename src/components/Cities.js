import PropTypes from 'prop-types';
import React, { Component } from 'react';
import City from './City';

class Cities extends Component {
	componentWillMount() {
		let cities = localStorage.getItem('storedCities');
		if (cities !== null && cities !== '') {
			cities = cities.split('&');
			this.props.updateStoredCities(cities);
		}
	}

	componentDidUpdate() {
		const stringCities = this.props.storedCities.join('&');
		localStorage.setItem('storedCities', stringCities);
	}

	cityList = () => this.props.storedCities.map((city, idx) => <City name={city} key={idx} fetchWeather={this.props.fetchWeather}/>);
	render() {
		return (
			<ul className="nav nav-pills nav-stacked flex-column">
				<li className="nav-header">Recent Cities <small>Â· <a href="#" onClick={this.props.clearStoredCities}>Clear</a></small></li>
				{this.cityList()}
			</ul>
		);
	}
};

Cities.defaultProps = {
	storedCities: [],
	fetchWeather: PropTypes.func,
};

Cities.propTypes = {
	storedCities: PropTypes.arrayOf(PropTypes.string),
	fetchWeather: PropTypes.func,
};

export default Cities;