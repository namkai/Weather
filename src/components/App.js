import React from 'react';
import '../style/App.css';
import SearchBar from './SearchBar';

export const App = () => (
  <div className="container">
    <SearchBar />
  </div>
);

export default App;
