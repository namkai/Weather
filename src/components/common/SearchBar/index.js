import { compose, pure, withHandlers, withState } from 'recompose';
import SearchBar from './view';

const addQueryState = withState('query', 'setQuery', '');

const addStateHandlers = withHandlers({
  updateQuery: ({ setQuery }) => ({ target: { value } }) => setQuery(() => value),
  onSearchForWeather: ({ query, setQuery, fetchWeather }) => (event) => {
    event.preventDefault();
    fetchWeather(query);
    setQuery(() => '');
  },
});

const PureSearchBar = pure(SearchBar);

export default compose(addQueryState, addStateHandlers)(PureSearchBar);
