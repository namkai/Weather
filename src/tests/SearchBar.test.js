import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { SearchBar } from '../components/SearchBar';

it('renders correctly', () => {
	const searchbar = shallow(
		<SearchBar >
			<strong>Hello World!</strong>
		</SearchBar>,
	);
	expect(toJson(searchbar)).toMatchSnapshot();
});
