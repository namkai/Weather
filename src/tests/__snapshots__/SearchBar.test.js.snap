// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-lg-6"
  >
    <div
      className="input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Search for..."
        type="text"
      />
      <span
        className="input-group-btn"
      >
        <button
          className="btn btn-default"
          onClick={[Function]}
          type="button"
        >
          Get Weather!
        </button>
      </span>
    </div>
  </div>
</div>
`;
